File: is52038b-labs/03/cpp/DynamicArrayTest.cpp
#include <cmath>

#include <cppunit/extensions/HelperMacros.h>
#include <cppunit/CompilerOutputter.h>
#include <cppunit/extensions/TestFactoryRegistry.h>
#include <cppunit/ui/text/TestRunner.h>

#include "DynamicArray.hpp"

class DynamicArrayTest : public CppUnit::TestFixture {
  CPPUNIT_TEST_SUITE(DynamicArrayTest);
  CPPUNIT_TEST(testConstructor);
  CPPUNIT_TEST(testSmall);
  CPPUNIT_TEST(testLargePush);
  CPPUNIT_TEST(testLargePop);
  CPPUNIT_TEST_SUITE_END();
public:
  void setUp() { };
  void tearDown() { };
  void testConstructor();
  void testSmall();
  void testLargePush();
  void testLargePop();
};

void DynamicArrayTest::testConstructor() {
  DynamicArray *a = new DynamicArray(10);
  CPPUNIT_ASSERT_EQUAL_MESSAGE("A fresh dynamic array should have zero length", ((size_t) 0UL), a->length());
  delete a;
}

void DynamicArrayTest::testSmall() {
  DynamicArray *a = new DynamicArray(10);
  a->push(1);
  CPPUNIT_ASSERT_EQUAL(((size_t) 1UL), a->length());
  CPPUNIT_ASSERT_EQUAL(1, a->select(0));
  a->push(3);
  CPPUNIT_ASSERT_EQUAL(((size_t) 2UL), a->length());
  CPPUNIT_ASSERT_EQUAL(1, a->select(0));
  CPPUNIT_ASSERT_EQUAL(3, a->select(1));
  a->store(2, 0UL);
  CPPUNIT_ASSERT_EQUAL(((size_t) 2UL), a->length());
  CPPUNIT_ASSERT_EQUAL(2, a->select(0));
  CPPUNIT_ASSERT_EQUAL(3, a->select(1));
  delete a;
}

void DynamicArrayTest::testLargePush() {
  DynamicArray *a = new DynamicArray(10);
  for (int i = 0; i < 100; i++) {
    a->push(i);
  }
  CPPUNIT_ASSERT_EQUAL(((size_t) 100UL), a->length());
  CPPUNIT_ASSERT_EQUAL(99, a->select(99));
  CPPUNIT_ASSERT_EQUAL(3, a->select(3));
  delete a;
}

void DynamicArrayTest::testLargePop() {
  DynamicArray *a = new DynamicArray(10);
  for (int i = 1; i <= 100; i++) {
    a->push(i);
  }
  for (int i = 100; i >= 1; i--) {
    CPPUNIT_ASSERT_EQUAL((size_t) i, a->length());
    CPPUNIT_ASSERT_EQUAL(i, a->pop());
    CPPUNIT_ASSERT_EQUAL_MESSAGE("pop should reduce the length of a dynamic array by 1", (size_t) (i-1), a->length());
  }
  delete a;
}

CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(DynamicArrayTest, "DynamicArray");
CPPUNIT_REGISTRY_ADD_TO_DEFAULT("DynamicArray");

/* FIXME: more boilerplate */
int main(int argc, char *argv[]) {
  if (argc > 2) {
    std::cerr << "Usage: " << argv[0] << " [suitename]" << std::endl;
    return 1;
  }
  CppUnit::Test *test;
  CppUnit::TextUi::TestRunner runner;
  if (argc == 2)
    test = CppUnit::TestFactoryRegistry::getRegistry(argv[1]).makeTest();
  else
    test = CppUnit::TestFactoryRegistry::getRegistry().makeTest();

  runner.addTest(test);
  runner.setOutputter(new CppUnit::CompilerOutputter(&runner.result(), std::cerr));

  bool wasSuccessful = runner.run();
  return wasSuccessful ? 0 : 1;
}

File: is52038b-labs/03/java/DynamicArrayTest.java
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertEquals;
import org.junit.Test;
import org.junit.Before;
import org.junit.runner.JUnitCore;
import org.junit.runner.Result;

public class DynamicArrayTest {
    @Test
    public void testConstructor() {
        DynamicArray a = new DynamicArray(10);
        assertEquals("A fresh dynamic array should have zero length", 0, a.length());
    }
    @Test
    public void testSmall() {
        DynamicArray a = new DynamicArray(10);
        a.push(1);
        assertEquals(1, a.length());
        assertEquals(1, a.select(0));
        a.push(3);
        assertEquals(2, a.length());
        assertEquals(1, a.select(0));
        assertEquals(3, a.select(1));
        a.store(2, 0);
        assertEquals(2, a.length());
        assertEquals(2, a.select(0));
        assertEquals(3, a.select(1));
    }
    @Test
    public void testLargePush() {
        DynamicArray a = new DynamicArray(10);
        for (int i = 0; i < 100; i++) {
            a.push(i);
        }
        assertEquals(100, a.length());
        assertEquals(99, a.select(99));
        assertEquals(3, a.select(3));
    }
    @Test
    public void testLargePop() {
        DynamicArray a = new DynamicArray(10);
        for (int i = 1; i <= 100; i++) {
            a.push(i);
        }
        for (int i = 100; i >= 1; i--) {
            assertEquals(i, a.length());
            assertEquals(i, a.pop());
            assertEquals("pop should reduce the length of a dynamic array by 1", i-1, a.length());
        }
    }
    public static void main(String args[]) {
        JUnitCore core = new JUnitCore();
        core.run(DynamicArrayTest.class);
    }
    
}
